pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -17,239 +17,230 @@
+      int bitmapOffset = (_cardinality + 1) * Integer.BYTES;
+      offsetDataStream.writeInt(bitmapOffset);
+      bitmapFileStream.getChannel().position(bitmapOffset);
+
+      int startIndex = 0;
+      for (int dictId = 0; dictId < _cardinality; dictId++) {
+        MutableRoaringBitmap bitmap = new MutableRoaringBitmap();
+        int endIndex = getInt(_invertedIndexLengthBuffer, dictId);
+        for (int i = startIndex; i < endIndex; i++) {
+          bitmap.add(getInt(_invertedIndexValueBuffer, i));
         }
-      }
-
-      // Append the bitmap data file to the inverted index file
-      try (FileChannel out = new FileOutputStream(_invertedIndexFile, true).getChannel();
-          FileChannel in = new FileInputStream(tempBitmapDataFile).getChannel()) {
-        // NOTE: For some reason, it seems like the second argument of transferFrom is relative on Linux while it is
-        // an absolute position on MacOS X. As such, we reposition the stream to 0 on both platforms to make it an
-        // absolute position call.
-        out.position(0).transferFrom(in, out.size(), in.size());
+        startIndex = endIndex;
+
+        // Write offset and bitmap into file
+        bitmapOffset += bitmap.serializedSizeInBytes();
+        // Check for int overflow
+        Preconditions.checkState(bitmapOffset > 0, "Inverted index file: %s exceeds 2GB limit", _invertedIndexFile);
+        offsetDataStream.writeInt(bitmapOffset);
+        bitmap.serialize(bitmapDataStream);
       }
     } catch (Exception e) {
       FileUtils.deleteQuietly(_invertedIndexFile);
       throw e;
-    } finally {
-      FileUtils.deleteQuietly(tempBitmapDataFile);
     }
   }
 
   @Override
-  public void close() {
-    MmapUtils.unloadByteBuffer(_valueByteBuffer);
-    MmapUtils.unloadByteBuffer(_numValuesByteBuffer);
-    MmapUtils.unloadByteBuffer(_postingListLengthByteBuffer);
-    MmapUtils.unloadByteBuffer(_postingListValueByteBuffer);
-    MmapUtils.unloadByteBuffer(_postingListCurrentOffsetByteBuffer);
+  public void close() throws IOException {
+    destroyBuffer(_forwardIndexValueBuffer, _forwardIndexValueBufferFile);
+    _forwardIndexValueBuffer = null;
+    destroyBuffer(_forwardIndexLengthBuffer, _forwardIndexLengthBufferFile);
+    _forwardIndexLengthBuffer = null;
+    destroyBuffer(_invertedIndexValueBuffer, _invertedIndexValueBufferFile);
+    _invertedIndexValueBuffer = null;
+    destroyBuffer(_invertedIndexLengthBuffer, _invertedIndexLengthBufferFile);
+    _invertedIndexLengthBuffer = null;
+  }
+
+  private static void putInt(LBufferAPI buffer, long index, int value) {
+    buffer.putInt(index << 2, value);
+  }
+
+  private static int getInt(LBufferAPI buffer, long index) {
+    return buffer.getInt(index << 2);
+  }
+
+  private LBufferAPI createBuffer(long size, File mmapFile) throws IOException {
+    if (_useMMapBuffer) {
+      return new MMapBuffer(mmapFile, 0L, size, MMapMode.READ_WRITE);
```

Comment:
this is nice, the only problem is that our metrics on memory allocated will now be off. I am not sure how to fix the problem other than using list of pinot data buffers

Reply:
Added a TODO to add support to trackable big size buffer

Reply:
i would feel more comfortable if you add a TODO here as well, since these APIs will also change when we make large bufferes trackable? thanks. We need to get that in fairly quickly, since we don't want to be allocating buffers without tracking them via metrics.